Sources/X509/OCSP/OCSPPolicy.swift

```diff
@@ -230,24 +268,95 @@ public struct OCSPVerifierPolicy<Requester: OCSPRequester>: VerifierPolicy, Send
                 return .failsToMeetPolicy(reason: "OCSP response does not include a response for the queried certificate \(requestedCertID) - responses: \(basicResponse.responseData.responses)")
             }
             
+            switch await validateResponseSignature(basicResponse, issuer: issuer) {
+            case .meetsPolicy:
+                break
+            case .failsToMeetPolicy(let reason):
+                return .failsToMeetPolicy(reason: reason)
+            }
+            
+            switch response.verifyTime(validationTime: self.validationTime) {
+            case .meetsPolicy:
+                break
+            case .failsToMeetPolicy(let reason):
+                return .failsToMeetPolicy(reason: reason)
+            }
+            
             switch response.certStatus {
-            case .good:
-                switch response.verifyTime(now: self.now ?? Date()) {
-                case .meetsPolicy:
-                    break
-                case .failsToMeetPolicy(let reason):
-                    return .failsToMeetPolicy(reason: reason)
-                }
-                
-                // TODO: verify signature: rdar://104687979
-                return .meetsPolicy
             case .revoked(let info):
                 return .failsToMeetPolicy(reason: "revoked through OCSP, reason: \(info.revocationReason?.description ?? "nil")")
             case .unknown:
                 return .failsToMeetPolicy(reason: "OCSP response returned as status unknown")
+            case .good:
+                return .meetsPolicy
             }
         }
     }
+    
+    private func validateResponseSignature(
+        _ basicResponse: BasicOCSPResponse,
+        issuer: Certificate
+    ) async -> PolicyEvaluationResult {
+        let responderID = basicResponse.responseData.responderID
+        let potentialCertificates = [issuer] + [basicResponse.certs?.first].compactMap { $0 }
+        let leafCertificate = potentialCertificates.first { certificate in
+            switch responderID {
+            case .byName(let subject):
+                return certificate.subject == subject
+            case .byKey(let subjectKeyIdentifier):
+                return (try? certificate.extensions.subjectKeyIdentifier?.keyIdentifier == subjectKeyIdentifier.bytes) ?? false
```

Comment:
It's not clear to me that the contents of the subject key identifier are relevant here: the key hash seems to be able to be used regardless of that extension.

Reply:
Do you mean we should hash the public key on demand if `extensions.subjectKeyIdentifier` is nil?

Reply:
Done